src/bin/tikv-ctl.rs

```diff
@@ -21,25 +21,415 @@ extern crate clap;
+        println!("total region size: {}", convert_gbmb(total_size as u64));
+    }
+
+    fn dump_region_info(&self, region: u64, skip_tombstone: bool) {
+        let r = self.get_region_info(region);
+        if skip_tombstone {
+            let region_state = r.region_local_state.as_ref();
+            if region_state.map_or(false, |s| s.get_state() == PeerState::Tombstone) {
+                return;
+            }
+        }
+        let region_state_key = keys::region_state_key(region);
+        let raft_state_key = keys::raft_state_key(region);
+        let apply_state_key = keys::apply_state_key(region);
+        println!("region id: {}", region);
+        println!("region state key: {}", escape(&region_state_key));
+        println!("region state: {:?}", r.region_local_state);
+        println!("raft state key: {}", escape(&raft_state_key));
+        println!("raft state: {:?}", r.raft_local_state);
+        println!("apply state key: {}", escape(&apply_state_key));
+        println!("apply state: {:?}", r.raft_apply_state);
+    }
+
+    fn dump_all_region_info(&self, skip_tombstone: bool) {
+        for region in self.get_all_meta_regions() {
+            self.dump_region_info(region, skip_tombstone);
+        }
+    }
+
+    fn dump_raft_log(&self, region: u64, index: u64) {
+        let idx_key = keys::raft_log_key(region, index);
+        println!("idx_key: {}", escape(&idx_key));
+        println!("region: {}", region);
+        println!("log index: {}", index);
+
+        let mut entry = self.get_raft_log(region, index);
+        let data = entry.take_data();
+        println!("entry {:?}", entry);
+        println!("msg len: {}", data.len());
+
+        let mut msg = RaftCmdRequest::new();
+        msg.merge_from_bytes(&data).unwrap();
+        println!("{:?}", msg);
+    }
+
+    fn dump_mvccs_infos(
+        &self,
+        from: Vec<u8>,
+        to: Option<Vec<u8>>,
+        limit: Option<u64>,
+        cfs: Vec<&str>,
+        start_ts: Option<u64>,
+        commit_ts: Option<u64>,
+    ) {
+        let to = to.unwrap_or_default();
+        let limit = limit.unwrap_or_default();
+        if to.is_empty() && limit == 0 {
+            eprintln!(r#"please pass "to" or "limit""#);
+            process::exit(-1);
+        }
+        if !to.is_empty() && to < from {
+            eprintln!("The region's from pos must greater than the to pos.");
+            process::exit(-1);
+        }
+        let scan_future = self.get_mvcc_infos(from, to, limit).for_each(
+            move |(key, mut mvcc)| {
+                println!("key: {}", escape(&key));
+                if cfs.contains(&CF_LOCK) && mvcc.has_lock() {
+                    let mut lock_info = mvcc.take_lock();
+                    if start_ts.map_or(true, |ts| lock_info.get_lock_version() == ts) {
+                        // FIXME: "lock type" is lost in kvproto.
+                        let pk = escape(lock_info.get_primary_lock()).into_bytes();
```

Comment:
Why `escape`?

Reply:
I think primary key points another valid data key in rocksdb? So we escape it in order to we can easily pass it into another tikv-ctl command, such as `print`.

Reply:
Output protobuf message in debug format will do that automatically.